File: sources/academy/webscraping/scraping_basics_javascript/exercises/eurozone_population.mjs
import * as cheerio from 'cheerio';

async function download(url) {
  const response = await fetch(url);
  if (!response.ok) {
    throw new Error(`HTTP ${response.status}`);
  }
  const html = await response.text();
  return cheerio.load(html);
}

function parsePopulation($) {
  for (const element of $('li').toArray()) {
    const text = $(element).text();
    if (text.includes('Population')) {
      const digits = text
        .replace('Population:', '')
        .replaceAll(' ', '');
      return Number.parseInt(digits, 10);
    }
  }
  throw new Error('Population not found');
}

const listingUrl = 'https://european-union.europa.eu/institutions-law-budget/euro/countries-using-euro_en';
const $ = await download(listingUrl);

const $euroCountriesAccordion = $('.ecl-accordion__item').first();
const $countryLinks = $euroCountriesAccordion.find('li a');

const promises = $countryLinks.toArray().map(async (element) => {
  const countryUrl = new URL($(element).attr('href'), listingUrl).href;
  const $country = await download(countryUrl);
  return parsePopulation($country);
});

const populations = await Promise.all(promises);
const totalPopulation = populations
  .filter((population) => Number.isInteger(population))
  .reduce((sum, population) => sum + population, 0);

console.log(totalPopulation);

File: sources/academy/webscraping/scraping_basics_python/exercises/eurozone_population.py
import httpx
from bs4 import BeautifulSoup
from urllib.parse import urljoin


def download(url: str) -> BeautifulSoup:
    response = httpx.get(url)
    response.raise_for_status()
    return BeautifulSoup(response.text, "html.parser")


def parse_population(country_soup: BeautifulSoup) -> int | None:
    for item in country_soup.select("li"):
        if "Population" in item.text:
            digits = item.text.replace("Population:", "").replace(" ", "")
            return int(digits)
    raise ValueError("Population not found")


listing_url = "https://european-union.europa.eu/institutions-law-budget/euro/countries-using-euro_en"
listing_soup = download(listing_url)

total_population = 0
euro_countries_accordion = listing_soup.select(".ecl-accordion__item")[0]
for country_link in euro_countries_accordion.select("li a"):
    country_url = urljoin(listing_url, country_link["href"])
    country_soup = download(country_url)
    total_population += parse_population(country_soup)
print(total_population)
